lib/fileset/default.nix

```diff
@@ -0,0 +1,580 @@
+    isAttrs
+    ;
+
+  inherit (lib.trivial)
+    min
+    ;
+
+  inherit (lib.attrsets)
+    mapAttrs
+    attrByPath
+    setAttrByPath
+    attrNames
+    filterAttrs
+    ;
+
+  inherit (lib.sources)
+    pathType
+    ;
+
+  inherit (lib.lists)
+    length
+    head
+    tail
+    elem
+    elemAt
+    take
+    commonPrefixLength
+    removePrefix
+    hasPrefix
+    foldl'
+    ;
+
+  inherit (lib.path.parts)
+    deconstruct
+    construct
+    ;
+
+  inherit (lib.strings)
+    splitString
+    substring;
+
+  # Internal file set structure:
+  #
+  # # A set of files
+  # <fileset> = {
+  #   _type = "fileset";
+  #
+  #   # TODO: Indicate that these properties are internal
+  #
+  #   # A tree representation of all included files under the subpath
+  #   tree = <entry>;
+  # };
+  #
+  # # A directory entry value
+  # <entry> =
+  #   # A nested directory
+  #   <directory>
+  #
+  #   # A nested file
+  #   | <file>
+  #
+  #   # A removed file or directory
+  #   # This is represented like this instead of removing the entry from the attribute set because:
+  #   # - It improves laziness
+  #   # - It allows keeping the attribute set as a `builtins.readDir` cache
+  #   | null
+  #
+  # # A directory
+  # <directory> =
+  #   # The inclusion state for every directory entry
+  #   { <name> = <entry>; }
+  #
+  #   # All files in a directory, recursively.
+  #   # Semantically this is equivalent to `builtins.readDir path`, but lazier, because
+  #   # operations that don't require the entry listing can avoid it.
+  #   # This string is chosen to be compatible with `builtins.readDir` for a simpler implementation
+  #   "directory";
+  #
+  # # A file
+  # <file> =
+  #   # A file with this filetype
+  #   # These strings match `builtins.readDir` for a simpler implementation
+  #   "regular" | "symlink" | "unknown"
+  # */
+
+
+  # TODO: Document limitation that directories that are fully filtered out won't be included. The only way to end up with an empty directory is to have a recursively included directory which is empty in the source tree
+
+  withoutExcluded = tree:
+    if isAttrs tree then
+      let
+        result = filterAttrs
+          (name: value: value != null)
+          (mapAttrs
+            (name: withoutExcluded)
+            tree
+          );
+      in
+        if result == {} then
+          null
+        else
+          result
+    else tree;
+
+  create = tree: {
+    _type = "fileset";
+    inherit tree;
+    # Computed attributes for caching purposes
+    # The tree without any null's and empty attribute sets
+    # Strict
+    # TODO: Still needed?
+    # FIXME: This makes equality comparison using `==` much more expensive, which would otherwise work without problems
+    # No, `==` won't ever work well, because we have the recursive directory cutoff thing, so let's not worry about that, `==` is not supported, maybe we should set `_equalityAlwaysReturnsFalse = x: x` because functions always compare false
```

Comment:
Perhaps better would be `_noEq = throw "filesets are not comparable";`, although that does defeat `deepSeq`. I'd prefer a dialect without broken equality over one with `deepSeq`.

Reply:
Fortunately the `_sparseTree` isn't needed after all, but the `==` problem is still there indeed. I now changed this to a `_noEval = throw "..."`, where the error message gives some alternatives for when equality or printing of file sets is needed